django_socio_grpc/request_transformer/grpc_socio_proxy_context.py

```diff
@@ -21,7 +21,7 @@ def __init__(self, grpc_context, grpc_action):
         grpc_request_metadata = self.convert_metadata_to_dict(
             grpc_context.invocation_metadata()
         )
-        self.headers = json.loads(grpc_request_metadata.get(self.HEADERS_KEY, "{}"))
+        self.headers = grpc_request_metadata # json.loads(grpc_request_metadata.get(self.HEADERS_KEY, "{}"))
```

Comment:
As we use request_metadata as headers and query params to keep REST library compatibility we need to specify what is headers and what is query params. 
Then grpc metadata is string format so we json load it to have a correct header dictionnary.

Please set your headers in a grpc metadata names HEADERS with header in json object format.

Reply:
There is a some bug. I have fixed it because it didn't work in gRPC authorization flow. I sent requests via Postman and also via Python client and  grpc_request_metada always was dict with headers like a 

{
'Authorization': 'Bearer xxxx',
'User-agent': 'blah....'
}

It didn't contain json. Are you sure that this piece of code is correct?

Reply:
Yes this part of code is correct.
Your metadata should look like:
{
 'HEADERS': '{"Authorization": "Bearer xxxx", "User-agent": "blah...."}'
}

The way to transform grpc metadata to HTTP request header is pure framework implementation not grpc one. If you are using a specific grpc authorisation flow I may want to see the documentation of it to try it and implement both way.

But if you are using django authentication and permissions api my recommendation should work.

Reply:
I use django auth and permissions and it didn't work with original code :(
I got incorrect data with request = SocioProxyHttpRequest class (no token) in my authentication class in
get_authorization_header(request) because in this file is probably bug with headers treatment. Please try to send real gRPC request with Authorization header and try to get it in [django_socio_grpc/request_transformer/grpc_socio_proxy_context.py]

Reply:
Huuuum I never tried postman in is new version supporting grpc. Can you try using postman and use a custom header named HEADERS an put the value '{"Authorization": "Bearer xxxx", "User-agent": "blah...."}' inside it ?

Indeed django-socio-grpc will not work with usual postman Authorisazion header. But this is in fact wanted. gRPC has metadata not header. We choose to make it retrocompatible by using the HEADERS metadata but this is juste an implementation.

We could make a settings to inject directly the metadata as headers. But then I am afraid FILTERS and PAGINATION metadata will be inject as headers and if name match some http header having some invisible issues.

Reply:
I understand that there is metada and no headers. But I have checked and usually authorization record is placing in the 1st level (like a header). So I believe that header with name HEADERS and JSON will work (I didn't realize it during testing today morning), but I don't understand why should we do it in such way?!I didn't find such approach in the network.

Reply:
It's just to separate different retrocompatible logic.
As you see in the code we are doing the same with PAGINATION and FILTERS. We wanted to avoid this to be merged with headers.
Of course we can think of something about ignoring this keys when injecting metadata into HEADERS.

I think this may be a django settings to let the user do what it prefer.

Don't know yet if better to have a HEADERS_METADATA_KEY settings and if None or empty using directly the metadata or a USE_METADATA_AS_HEADERS boolean to be more explicit.

Reply:
For me it is not problem to change authorisation for client, but question is what is correct way for gRPC standards / approaches?! 

Maybe having it in settings is good idea. I can change it as my branch doesn't pass all tests right now due to this change.

Reply:
Actually as there is no Authorisation library on python with grpc we are making the standards ourself with the different tools developper are using.
When we started this project postman was not supporting grpc we used bloomrpc that take a json object as metadata.
Now that postman support gRPC and seem to say that the authorization key should be a matadata named authorization we may change the default behavior to adapt to it.

Reply:
To be completely honest there is standard for authorization in grpc: https://grpc.io/docs/guides/auth/#python
But there are setted in the channel and not in the request metadata. 
It was hard to implement so we choose to use metadata.

Reply:
OK, no problem, I can support any of these - change it back or introduce settings. Please let me know what you prefer.

Reply:
I will go with the settings then.
If possible having a setting named MAP_METADATA_KEYS with default:
```
MAP_METADATA_KEYS = {"HEADERS": "HEADERS", "PAGINATION": "PAGINATION", "FILTERS": "FILTERS"}
```
And add a method in SocioProxyHttpRequest:
```
get_from_metadata(self, metadata_keys):
  metadata_key = MAP_METADATA_KEYS.get(metadata_keys, None)
  if not metadata_key: 
    return self.grpc_request_metadata # think about putting grpc_request_metadata in self in the __init__ function
  return json.loads(grpc_request_metadata.get(metadata_key, "{}"))
``` 
and call this methods instead of doing the json.load the three time it's called inside SocioProxyHttpRequest